### ISSUE TITLE
Parse IPv6 zone identifiers

### ISSUE THREAD

**Issue by famfo** [rust-lang/libs-team#476]
# Proposal

Extend the `Ipv6Addr` struct with support for zone identifiers (interface indexes and interface names) including string parsing.

## Problem statement

Ipv6 addresses have zone identifiers (`<address>%<zone_id>`) for e.g. link-local and multicast addresses as per RFC4007. This is currently not supported by the `core` and `std` implementations of an `Ipv6Addr`.

## Motivating examples or use cases

Use cases include addressing multicast IP addresses on specific interfaces for e.g. wake on LAN when the default route does not exit into the local network where devices are located or neighbor discovery over multicast on specific interfaces for dynamic routing protocols.

Other expected uses actively break when trying to parse a file that contains an Ipv6 address scoped using an interface name (e.g. hickory DNS and `/etc/resolv.conf` containing `nameserver fe80::2a0:57ff:fe6f:42ee%wlp3s0`).

## Solution sketch

- move Ipv6 address zone identifier from `SocketAddrV6` to `Ipv6Addr`
- add `libc::if_nametoindex` wrappers to platform specific APIs in `std`
- add the ability to extend the `core::net::parser` Ipv6 parser with custom zone identifier parsers
- extend the Ipv6 parser to resolve zone identifier by name in `std`

## Alternatives

- only adding interface name resolution to `SocketAddrV6` in `std`
  - cons: this will not work when parsing IP addresses, only socket addresses

## Links and related work

<https://datatracker.ietf.org/doc/html/rfc3513>
<https://datatracker.ietf.org/doc/html/rfc4007>
<https://datatracker.ietf.org/doc/html/rfc3493>
<https://github.com/rust-lang/rust/issues/65976>

## What happens now?

This issue contains an API change proposal (or ACP) and is part of the libs-api team [feature lifecycle]. Once this issue is filed, the libs-api team will review open proposals as capability becomes available. Current response times do not have a clear estimate, but may be up to several months.

[feature lifecycle]: https://std-dev-guide.rust-lang.org/development/feature-lifecycle.html

## Possible responses

The libs team may respond in various different ways. First, the team will consider the *problem* (this doesn't require any concrete solution or alternatives to have been proposed):

- We think this problem seems worth solving, and the standard library might be the right place to solve it.
- We think that this probably doesn't belong in the standard library.

Second, if there's a concrete solution:

- We think this specific solution looks roughly right, approved, you or someone else should implement this. (Further review will still happen on the subsequent implementation PR.)
- We're not sure this is the right solution, and the alternatives or other materials don't give us enough information to be sure about that. Here are some questions we have that aren't answered, or rough ideas about alternatives we'd want to see discussed.

**Comment by kennytm** [comment]
I think the `<core::net::Ipv6Addr>::from_str()` impl should remain platform-agnostic and support just the numerical form `fe80::abcd:1234%2`. It is similar to that you can't parse `127.0.0.1:tcp` into a SocketAddr.

**Comment by famfo** [comment]
@scottmcm

> that it could just be an unsigned integer of some kind? Though your example doesn't work with that.

Using an unsigned integer would work if converting interface names to interface indexes as @kennytm already noted.

> But even then, it no longer being just 128-bit feels like it'd be a surprise to a bunch of people, and potentially bad.
>
> TL/DR it feels like this should be a different type, not part of Ipv6Addr.

A scope identifier is part of an IPv6 address and I feel it more surprising that valid IPv6 addresses are not properly parsed. I don't see how a different type would solve the issue properly.

---

@kennytm 

> I think the <core::net::Ipv6Addr>::from_str() impl should remain platform-agnostic and support just the numerical form fe80::abcd:1234%2. It is similar to that you can't parse 127.0.0.1:ssh into a SocketAddr even if you know it is port 22 through getservbyname.

Parsing interface names in `core` doesn't really make sense, that's why I proposed making the parser extendable for `std` (although I don't think `getservbyname` and `if_nametoindex` are a fair comparison).

**Comment by kennytm** [comment]
> Parsing interface names in `core` doesn't really make sense, that's why I proposed making the parser extendable for `std` 

I don't think it makes sense to produce a different answer for `"fe80::1234%eth0".parse::<Ipv6Addr>()` depending on whether the program is `#![no_std]`.

**Comment by famfo** [comment]
That is a fair argument. I think this could be solved with an extention trait, similar to `[ToSocketAddrs](https://doc.rust-lang.org/stable/std/net/trait.ToSocketAddrs.html#)`

**Comment by kennytm** [comment]
Oh also, while the zone ID is required to disambiguate non-global IPv6 address, it is never used in the wire format. An IPv6 packet does not contain the `scope_id` at all. 

So I think it does make sense that `core::net::Ipv6Addr` corresponds just to C's `struct in6_addr`, and `core::net::SocketAddrV6` is the entire `struct sockaddr_in6` where the `scope_id` lives.

**Comment by famfo** [comment]
> Oh also, while the zone ID is required to disambiguate non-global IPv6 address, it is never sent on the wire. An IPv6 packet does not contain the scope_id at all.

Not send on the wire but parsed before by e.g. the Linux kernel to determine on which interface to send it out.

> If I look at a [X-Forwarded-For](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-For) header, passing along a %3 because the proxy happened to have wan on its 3rd interface is irrelevant and not something I would want included.

Bad example in my opinion, if the peer is a link-local from the 2. interface it is a very important information. The zone identifier is explicitly passed with link-locals and multicast addresses, otherwise it is implicitly 0.

If proper Ipv6 parsing is only added to `SockAddressV6` passing only IP addresses without a port is not possible e.g. for user input to a program and Ipv6 addresses have to be passed as `[fe80::1%eth0]:0`

**Comment by kennytm** [comment]
> If proper Ipv6 parsing is only added to `SockAddressV6` passing only IP addresses without a port is not possible e.g. for user input to a program and Ipv6 addresses have to be passed as `[fe80::1%eth0]:0`

No this works:

```rust
use std::net::ToSocketAddrs as _;

fn main() {
    let socket_addr = ("fe80::1%3", 0).to_socket_addrs().unwrap().next().unwrap();
    dbg!(socket_addr);
    // socket_addr = [fe80::1%3]:0
}
```

**Comment by tliron** [comment]
There might be a related problem:

It seems that currently (Rust 1.86.0) even specifying a u32 zone in `SocketAddrV6` is not used when binding a listener. For example, specifying a zone that doesn't exist on the system still successfully binds to the address. I believe this should result in an error. And, of course, it should bind to the correct interface when specified.

As far as I understand it, the whole point of specifying zones is for binding listeners. At least that's how it works in APIs in other languages.

So, the problem is not just in parsing and representation (we definitely need to support strings!), but in the actual use of this feature.

**Comment by kennytm** [comment]
> It seems that currently (Rust 1.86.0) even specifying a u32 scope ID in `SocketAddrV6` is not used when binding a listener. For example, specifying a scope ID that doesn't exist on the system still successfully binds to the address. I believe this should result in an error. And, of course, it should bind to the correct interface when specified.
> 
> As far as I understand it, the whole point of specifying zones is for binding listeners. At least that's how it works in APIs in other languages.

This behavior is just reflecting how the OS works? The scope ID is only used if you have supplied a link-local address (`fe80::/10`), if your listening address is not of this class (e.g. if you listen to `::`) the scope ID will be ignored.

**Comment by famfo** [comment]
I think there are two options to move forward with this proposal:

- the one with backwards compatibility: an extension to `std::net::ToSocketAddrs` resolving the zone ID from a given interface name on supported libcs
- the more "correct" option, which would be a breaking change: moving the zone ID to `std::net::Ipv6Addr`

Personally: I prefer the second option for correctness, for example in routing scenarios where next-hops are only an IP address selecting the interface traffic is being sent out over (making zone IDs very important, even when not being sent over the wire). I also acknowledge that breaking changes are very undesirable, even for correctness and that makes the option very unappealing.

**Comment by CutestNekoAqua** [comment]
i agree with famfo here and i just faced that issue when trying to work with IPv6 link-locals. I think this needs to be a breaking change but i guess for that we need to wait for a new rust edition?

**Comment by kennytm** [comment]
So I've checked the prior art for this, and the population of having the scope-id included into their equivalent `Ipv6Addr` is pretty much 50/50.

| Library | `SocketAddrV6` equivalent | `Ipv6Addr` equivalent | Scope-ID in `Ipv6Addr`? | 
|---|---|---|---|
| Rust| `std::net::SocketAddrV6` | `std::net::Ipv6Addr` | ❌: `scope_id()` in `SocketAddrV6` |
| POSIX `<netinet/in.h>` | `struct sockaddr_in6` | `struct in6_addr` | ❌: `sin6_scope_id` in `sockaddr_in6` |
| [C++ P3482R1](https://wg21.link/p3482r1) | N/A | `std::net::ip::address_v6` | ✅: `scope_id()` in `ip::address_v6` |
| Python | (str, int, int, int) | N/A | ❌ |
| Java | [`java.net.InetSocketAddress`](https://docs.oracle.com/javase/8/docs/api/java/net/InetSocketAddress.html) | [`java.net.Inet6Address`](https://docs.oracle.com/javase/8/docs/api/java/net/Inet6Address.html) | ✅: `getScopeId()` in `Inet6Address` |
| Node.js | [`net.SocketAddress`](https://nodejs.org/api/net.html#class-netsocketaddress) | string | ✅: scope-ID can be parsed from the "string" |
| C# | [`System.Net.IPEndPoint`](https://learn.microsoft.com/en-us/dotnet/api/system.net.ipendpoint?view=net-9.0) | [`System.Net.IPAddress`](https://learn.microsoft.com/en-us/dotnet/api/system.net.ipaddress?view=net-9.0) | ✅: `ScopeId` in `IPAddress`
| Go (legacy) | [interface `net.Addr`](https://pkg.go.dev/net#Addr), impl:<br>&bull; [struct `net.TCPAddr`](https://pkg.go.dev/net#TCPAddr)<br>&bull; [struct `net.UDPAddr`](https://pkg.go.dev/net#UDPAddr)<br>&bull; [struct `net.IPAddr`](https://pkg.go.dev/net#IPAddr) | [type `net.IP`](https://pkg.go.dev/net#IP) | ❌: `Zone` in the various `*Addr` structures |
| Go 1.18 | [`netip.AddrPort`](https://pkg.go.dev/net/netip#AddrPort) | [`netip.Addr`](https://pkg.go.dev/net/netip#Addr) | ✅: `Zone()` in `Addr` |
| Ruby | [`Addrinfo`](https://docs.ruby-lang.org/en/3.4/Addrinfo.html) | [`IPAddr`](https://docs.ruby-lang.org/en/3.4/IPAddr.html) | ✅: `zone_id` in `IPAddr` |
| Swift Network | [`NWEndpoint`](https://developer.apple.com/documentation/network/nwendpoint) | [`IPv6Address`](https://developer.apple.com/documentation/network/ipv6address)<br>[`NWEndpoint.Host`](https://developer.apple.com/documentation/network/nwendpoint/host) | 🤔: `interface` in `NWEndpoint.Host` only (equivalent to Rust's `IpAddr`) |
| Dart | N/A | [InternetAddress](https://api.dart.dev/dart-io/InternetAddress/InternetAddress.html) | ✅: scope-ID is [stored in `InternetAddress`](https://github.com/dart-lang/sdk/blob/79cbfa032d220c5f7f9cb3753a69d6824bf1f78d/sdk/lib/_internal/vm/bin/socket_patch.dart#L176), but hidden from public API |
| Erlang/OTP<br>Elixir | N/A | [`inet:ip6_address/0`](https://www.erlang.org/doc/apps/kernel/inet.html#t:ip6_address/0) | ❌: `inet` module has no scope-id support, [one must use `socket:sockaddr_in6/0` instead](https://www.erlang.org/doc/apps/kernel/gen_tcp.html#connect/3-with-argument-sockaddr-since-otp-24-3). |
| Zig | [`std.net.Ip6Address`](https://ziglang.org/documentation/master/std/#std.net.Ip6Address) | N/A | ❌: `.sa.scope_id` in `Ip6Address` |
| D | [`std.socket.Internet6Address`](https://dlang.org/phobos/std_socket.html#.Internet6Address) | N/A | ❌: `Internet6Address` encapsulates an `sockaddr_in6` so one can get the `sin6_scope_id` through `.name()` + type casting |

Standard libraries that map 1:1 with the BSD socket library tend to put the scope-id into `SocketAddrV6`, and Rust was certainly in this camp (prior to rust-lang/rfcs#2832 these address types wrap the equivalent system struct).

**Comment by tliron** [comment]
> and `udp_socket.connect(("fe80::2a0:57ff:fe6f:42ee%wlp3s0", 53))` is correct enough.

This note surprised me, so I did some testing and indeed when you parse a socket address from a string it seems that a *zone* lookup happens and the correct *scope ID* is selected. So it seems socket zone *is* supported, albeit in a rather indirect manner. This surprised me because there's no explicit API for setting or looking up a zone. [`SocketAddrV6::new`](https://doc.rust-lang.org/std/net/struct.SocketAddrV6.html#method.new) only accepts a scope ID, and nowhere is zone lookup mentioned.

Have we failed to document this feature, or is there something I'm missing here?

**Comment by kennytm** [comment]
`ToSocketAddrs` with string input [just calls the system](https://github.com/rust-lang/rust/blob/ebd8557637b33cc09b6ee8273f3154d5d3af6a15/library/std/src/sys/net/connection/socket.rs#L319)'s `getaddrinfo(3)`, so it is the system calling `if_nametoindex` internally (after all this method is supposed to do DNS query converting `"example.com"` to "23.215.0.136" etc for instance, so it doing the comparably cheaper `if_nametoindex` too is understandable).

This treatment is guaranteed by [OpenBSD](https://man.openbsd.org/getaddrinfo.3#:~:text=The%20scope%20identifier%20is%20hardcoded%20to%20the%20name,.) and macOS. On Linux there is a [brief note](https://www.man7.org/linux/man-pages/man3/getaddrinfo.3.html#NOTES) and I tested it does work. Windows only allow numeric scope index. The overall support seems to be [taken for granted by `draft-ietf-6man-zone-ui-00`](https://www.ietf.org/archive/id/draft-ietf-6man-zone-ui-00.html).

I suppose the Rust documentation can be expanded to specify that parsing `"fe80::1%eth0"` is supported on all major targets. Note that there are targets which has no `getaddrinfo(3)`, e.g. SGX, UEFI, WASIp1, and in those cases even the numeric form `("fe80::1%3", 0).to_socket_addrs()` will throw "unsupported".

**Comment by tliron** [comment]
@kennytm Thanks for the explanation, I suspected that was the case. However, I hope we can agree this is an awkward situation.

Here's an example from my code, where `address` is `Ipv6Addr`. As you can see, to initialize `SocketAddrV6` I have to convert `address` to a string, concatenate the zone, and then call `to_socket_addrs`, which parses it back to an address, and then iterate it:

```rust
match &zone {
    Some(zone) => {
        // We can only set the zone by re-parsing the address
        (address.to_string() + "%" + zone, port)
            .to_socket_addrs()?
            .next()
            .ok_or(io::Error::other("no socket address found"))?
    }

    None => SocketAddrV6::new(address, port, flowinfo, 0).into(),
}
```

The difference between with-a-zone and without-a-zone is stark. I guess this works, but it seems very roundabout and not obvious to me.

It seems to me that there should be a version of `SocketAddrV6::new` that accepts zone. Perhaps it can be `SocketAddrV6::new_with_zone`? Underneath it could be exactly my code above, but at least it would be discoverable and ergonomic for users. With such a new API my code would be much more coherent:

```rust
match &zone {
    Some(zone) => SocketAddrV6::new_with_zone(address, port, flowinfo, zone).into(),
    None => SocketAddrV6::new(address, port, flowinfo, 0).into(),
}
```

**Comment by kennytm** [comment]
@tliron it seems to me that what you want is just a wrapper of `if_nametoindex`, something like:

```rust
// mod std::net
pub fn parse_scope_id(name: &str) -> std::io::Result<u32> {
    if let Ok(number) = name.parse() {
        Ok(number)
    } else {
        if_nametoindex(name)
    }
}
```

so you end up with:

```rust
let scope_id = if let Some(zone) = &zone {
   parse_scope_id(zone)?
} else {
   0
};
SocketAddrV6::new(address, port, flowinfo, scope_id)
```

that would be a distinct ACP though.

**Comment by tliron** [comment]
@kennytm Is that what I want? :) I just want a better and clearer experience for all Rust users. :)

I guess your proposed API would work if we can find cross-platform equivalents. I'm not sure about Windows. But why would it be distinct from this issue, which is titled "Parse IPv6 zone identifiers"?

Anyway, I feel there is some subtle documentation slippage between these terms (not just in Rust, but also in the sources Rust references). We have some links in the issue description, but it's useful to make these explicit.

"Scope" is defined in [RFC 4007](https://datatracker.ietf.org/doc/html/rfc4007) (2005), which to me seems the most recent and most authoritative. There is no mention there of "scope IDs". Rather, "zones" are introduced as *applying* to scopes. What we have are "zone indices". Relevant section:

> At present, there is no way for a node to automatically determine
   which of its interfaces belong to the same zones; e.g., the same link
   or the same multicast scope zone larger than interface.  In the
   future, protocols may be developed to determine that information.  In
   the absence of such protocols, an implementation must provide a means
   for manual assignment and/or reassignment of zone indices.
   Furthermore, to avoid performing manual configuration in most cases,
   an implementation should, by default, initially assign zone indices
   only as follows:
> * A unique interface index for each interface.
> * A unique link index for each interface.

And then:

> As already mentioned, to specify an IPv6 non-global address without ambiguity, an intended scope zone should be specified as well.  As a common notation to specify the scope zone, an implementation SHOULD support the following format: `<address>%<zone_id>`

...

> An implementation MAY support other kinds of non-null strings as <zone_id>.  However, the strings must not conflict with the delimiter character.  The precise format and semantics of additional strings is implementation dependent. One possible candidate for these strings would be interface names, ...

Rust's `SocketAddrV6` references [RFC 2553](https://datatracker.ietf.org/doc/html/rfc2553) (1999), which indeed mentions "scope_id" as a u32 and never mentions "zones". Note that this RFC is specifically recommendations about *an API* and provides examples for BSD 4.3. Relevant section:

> IPv6 addresses are scoped [[2](https://datatracker.ietf.org/doc/html/rfc2553#ref-2)] so they could be link-local, site, organization, global, or other scopes at this time undefined.  To support applications that want to be able to identify a set of interfaces for a specific scope, the IPv6 sockaddr_in structure must support a field that can be used by an implementation to identify a set of interfaces identifying the scope for an IPv6 address.

To summarize the various slippages as I understand them:

1. What the "zone index" (or "scope ID") means depends on the implementation (the local networking stack) 
2. It is *often* defined in implementation APIs as a u32 0-based index of the local network interfaces. Is this the case in all Rust-supported platforms?
3. When represented as strings, the "zone index" (or "scope ID") after the `%` *could* be, well, anything (within character restrictions). It is, again, implementation specific. In Linux it *can* be the interface name, which is mapped to an interface index via [if_nametoindex](https://man7.org/linux/man-pages/man3/if_nametoindex.3.html). 

Also see a good superuser summary [here](https://superuser.com/a/1069088) (2016) that discusses some platform differences. 

I'm left unsure as what would be best for a cross-platform API in Rust. What I *am* sure is that `to_socket_addrs` and `SocketAddrV6::new` should be functionally equivalent. Right now it's very confusing that (only on Linux) they use different code paths with different features and results.

**Comment by kennytm** [comment]
@tliron 

> But why would it be a distinct ACP from this issue, which is titled "Parse IPv6 zone identifiers"?

it is a distinct ACP because this one is talking about moving `scope_id` into `Ipv6Addr`. If one just wants a `parse_scope_id`-equivalent function, the content of the entire ACP should be rewritten.

> What I _am_ sure is that `to_socket_addrs` and `SocketAddrV6::new` should be functionally equivalent.

It is *definitely* a wrong assessment because `to_socket_addrs` will do DNS query. `SocketAddrV6::new` does not take any string argument either, and it being a `const fn` meaning it can never do `if_nametoindex`.

**Comment by tliron** [comment]
> It is _definitely_ a wrong assessment because `to_socket_addrs` will do DNS query.

Oops, yes, they are quite different. Well, my point is specifically about the setting-the-zone feature. That functionality, I *am* sure, should be prominently available in `SocketAddrV6`.

> `SocketAddrV6::new` does not take any string argument either, and it being a `const fn` meaning it can never do `if_nametoindex`.

Yes, that is why I suggested a new (non-const) API function (`new_with_zone`, maybe?), one that furthermore returns an error if the zone lookup fails.

**Comment by famfo** [comment]
> I think a biggest problem is [how do you define Step for Ipv6Addr](https://doc.rust-lang.org/std/net/struct.Ipv6Addr.html#impl-Step-for-Ipv6Addr): how would you step from fe80::1%3 to fe80::2%2?

I don't think that `Step` should change the interface scope of an IP address at all. If I were to create a range on an IPv6 address I'd expect all of the addresses to end up in the same scope, be it the global scope or a specific interface.

**Comment by kennytm** [comment]
@famfo I suppose then you want `fe80::1%3` unable to reach `fe80::2%2` no matter how many times `forward_checked` or `backward_checked` is called? 

This means the `Ipv6Addr` type got split into 2<sup>32</sup> independently traversable strands. This would make `Ipv6Addr` the first Step type that behaves like $\omega \cdot 2^{32}$, while all the remaining like `i128` or `char` or `Ipv4Addr` is just a single strand like $\omega$. 
While not currently prohibited by Step, I'm not sure if there is any hidden assumption that if `T: Step` that all valid values of `T` can be covered through backward/forward movements.

Additionally, because of the invariants spelled out in Step, this forces some PartialOrd behavior. Let's say `d7 == fe80::7%4`, `e8 == fe80::8%5`, `d9 == fe80::9%4`. We want both of the following:
1. `d7.forward(2) == b9`
2. `e8` is unreachable from `d7, d9` by any means (you have to go forward infinite steps from `d7` to reach `e8`)

This means we can't have `d7 < e8 < d9`. Because `Ipv6Addr` implemented Ord, `e9` can't be unordered against `d7, d9` either. So that forces either `e8 < d7 < d9` or `d7 < d9 < e8` and the naturally we prefer the latter because 4 < 5. 
In the other words the Ord for `Ipv6Addr` compares like `(self.scope_id(), self.to_bits())`[^1]. 

[^1]: Note that this ordering differs from `SockAddrV6` which is `#[derive]`d and compares like `(self.ip().to_bits(), self.port(), self.flowinfo(), self.scope_id())`.
